README.md
<p align="center">
  <a href="http://nestjs.com/" target="blank"><img src="https://nestjs.com/img/logo_text.svg" width="320" alt="Nest Logo" /></a>
</p>

[circleci-image]: https://img.shields.io/circleci/build/github/nestjs/nest/master?token=abc123def456
[circleci-url]: https://circleci.com/gh/nestjs/nest

  <p align="center">A progressive <a href="http://nodejs.org" target="_blank">Node.js</a> framework for building efficient and scalable server-side applications.</p>
    <p align="center">
<a href="https://www.npmjs.com/~nestjscore" target="_blank"><img src="https://img.shields.io/npm/v/@nestjs/core.svg" alt="NPM Version" /></a>
<a href="https://www.npmjs.com/~nestjscore" target="_blank"><img src="https://img.shields.io/npm/l/@nestjs/core.svg" alt="Package License" /></a>
<a href="https://www.npmjs.com/~nestjscore" target="_blank"><img src="https://img.shields.io/npm/dm/@nestjs/common.svg" alt="NPM Downloads" /></a>
<a href="https://circleci.com/gh/nestjs/nest" target="_blank"><img src="https://img.shields.io/circleci/build/github/nestjs/nest/master" alt="CircleCI" /></a>
<a href="https://coveralls.io/github/nestjs/nest?branch=master" target="_blank"><img src="https://coveralls.io/repos/github/nestjs/nest/badge.svg?branch=master#9" alt="Coverage" /></a>
<a href="https://discord.gg/G7Qnnhy" target="_blank"><img src="https://img.shields.io/badge/discord-online-brightgreen.svg" alt="Discord"/></a>
<a href="https://opencollective.com/nest#backer" target="_blank"><img src="https://opencollective.com/nest/backers/badge.svg" alt="Backers on Open Collective" /></a>
<a href="https://opencollective.com/nest#sponsor" target="_blank"><img src="https://opencollective.com/nest/sponsors/badge.svg" alt="Sponsors on Open Collective" /></a>
  <a href="https://paypal.me/kamilmysliwiec" target="_blank"><img src="https://img.shields.io/badge/Donate-PayPal-ff3f59.svg"/></a>
    <a href="https://opencollective.com/nest#sponsor"  target="_blank"><img src="https://img.shields.io/badge/Support%20us-Open%20Collective-41B883.svg" alt="Support us"></a>
  <a href="https://twitter.com/nestframework" target="_blank"><img src="https://img.shields.io/twitter/follow/nestframework.svg?style=social&label=Follow"></a>
</p>
  <!--[![Backers on Open Collective](https://opencollective.com/nest/backers/badge.svg)](https://opencollective.com/nest#backer)
  [![Sponsors on Open Collective](https://opencollective.com/nest/sponsors/badge.svg)](https://opencollective.com/nest#sponsor)-->

## Description

[Nest](https://github.com/nestjs/nest) framework TypeScript starter repository.

## Installation

```bash
$ npm install
```

## Running the app

```bash
# development
$ npm run start

# watch mode
$ npm run start:dev

# production mode
$ npm run start:prod
```

## API doc

```
http://localhost:3000/api/
```

## Test

```bash
# unit tests
$ npm run test

# e2e tests
$ npm run test:e2e

# e2e tests in docker
$ npm run docker-compose:test:e2e

# test coverage
$ npm run test:cov
```

## Support

Nest is an MIT-licensed open source project. It can grow thanks to the sponsors and support by the amazing backers. If you'd like to join them, please [read more here](https://docs.nestjs.com/support).

## Stay in touch

- Author - [Kamil Myśliwiec](https://kamilmysliwiec.com)
- Website - [https://nestjs.com](https://nestjs.com/)
- Twitter - [@nestframework](https://twitter.com/nestframework)

## License

Nest is [MIT licensed](LICENSE).


CHANGELOG.md
## [1.2.1](https://github.com/jeremy60607/nestjs-exchange-currencies/compare/v1.2.0...v1.2.1) (2021-12-05)


### Bug Fixes

* **deps:** pin dependency mathjs to 10.0.0 ([c88babd](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/c88babdc46461d57a8e2e7a5f995edf29b8fdf50))

# [1.2.0](https://github.com/jeremy60607/nestjs-exchange-currencies/compare/v1.1.0...v1.2.0) (2021-12-05)


### Bug Fixes

* add package-lock.json file ([c9e3bbd](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/c9e3bbde991bf67e503bfa5ecf4674ad90244777))
* add unit test and e2e test ([14f1d1d](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/14f1d1da72ab68e7d414d2a8f3780354f24baf50))
* **deps:** pin dependencies ([f40a09e](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/f40a09e111a67670745a126d07687bdf2db56c99))


### Features

* add currencies module ([02280d8](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/02280d87d11fd602534e00243e90748f14317b17))
* add exception const ([d2770c1](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/d2770c116aea1325e2c7d7e621fc7a74d380fbf7))
* implementation api endpoint ([709283d](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/709283d625202ccfeccb860af9500c6965e82c8d))
* implementation service function ([213ebff](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/213ebff760661cc86ef2160126c17d660fcfc81e))

# [1.1.0](https://github.com/jeremy60607/nestjs-exchange-currencies/compare/v1.0.0...v1.1.0) (2021-12-05)


### Features

* swagger for doc ([61809d8](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/61809d8dd34ee364b4a9172593b95fb741139c3f))

# 1.0.0 (2021-12-05)


### Bug Fixes

* fix semantic-release ([cd76bc5](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/cd76bc587dbd49f62a013270ae3826398a26de75))
* fix semantic-release in husky ([12d0ad1](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/12d0ad1285ae008edce4ac16f3eb12601149e288))
* modify releaserc ([b23344c](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/b23344c905f5661973d9881d023c2ee9565bbcc9))


### Features

* add conventional commit plugins ([56a4027](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/56a4027fa37df54bd2b593fd053ee20c9a313e9d))
* add semantic-release when github action ([6563268](https://github.com/jeremy60607/nestjs-exchange-currencies/commit/656326802247b4e3d7bc3ac75f5813420854e8bb))


test/app.e2e-spec.ts
import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from './../src/app.module';

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        transform: true,
      }),
    );
    await app.init();
  });

  it('/ (GET)', () => {
    return request(app.getHttpServer())
      .get('/')
      .expect(200)
      .expect('Hello World!');
  });
});


test/currencies.e2e-spec.ts
import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication, ValidationPipe } from '@nestjs/common';
import * as request from 'supertest';
import { AppModule } from './../src/app.module';
import { CurrenciesModule } from '../src/currencies/currencies.module';
import { ExchangeCurrencyQueryDTO } from '../src/currencies/dto/currency.dto';
import { ExchangeCurrencyAO } from '../src/currencies/ao/currency.ao';

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeEach(async () => {
    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [CurrenciesModule],
    }).compile();

    app = moduleFixture.createNestApplication();
    app.useGlobalPipes(
      new ValidationPipe({
        transform: true,
      }),
    );
    await app.init();
  });

  it('/v1/currencies/exchange', () => {
    const exchangeCurrencyQuery: ExchangeCurrencyQueryDTO = {
      inputCurrency: 'TWD',
      targetCurrency: 'USD',
      value: 31555125.8456,
    };

    const exchangeCurrency: ExchangeCurrencyAO = {
      targetCurrency: 'USD',
      targetRate: 0.03281,
      value: '1,035,323.68',
    };

    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query(exchangeCurrencyQuery)
      .expect(200)
      .expect(exchangeCurrency);
  });

  it('check value negative', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        inputCurrency: 'USD',
        targetCurrency: 123,
        value: -31555125.8456,
      })
      .expect(400);
  });

  it('check value is number', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        inputCurrency: 'USD',
        targetCurrency: 123,
        value: '31555125.8456',
      })
      .expect(400);
  });

  it('check targetCurrency is not string', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        inputCurrency: 'USD',
        targetCurrency: 123,
        value: 31555125.8456,
      })
      .expect(400);
  });

  it('check inputCurrency is not string', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        inputCurrency: 123,
        targetCurrency: 'USD',
        value: 31555125.8456,
      })
      .expect(400);
  });

  it('check inputCurrency is empty', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        targetCurrency: 'USD',
        value: 31555125.8456,
      })
      .expect(400);
  });

  it('check targetCurrency is empty', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        inputCurrency: 'USD',
        value: 31555125.8456,
      })
      .expect(400);
  });

  it('check value is empty', () => {
    return request(app.getHttpServer())
      .get('/v1/currencies/exchange')
      .query({
        targetCurrency: 'USD',
        inputCurrency: 'USD',
      })
      .expect(400);
  });
});


src/currencies/currencies.service.spec.ts
import { Test, TestingModule } from '@nestjs/testing';
import { CurrenciesService } from './currencies.service';
import { ExchangeCurrencyQueryDTO } from './dto/currency.dto';
import { BadRequestException } from '@nestjs/common';
import { ExchangeRatePO } from './po/currency.po';
import { ExceptionConst } from '../const/exception.const';

const { COUNTRY_IS_INVALID, VALUE_IS_INVALID } = ExceptionConst;

describe('CurrenciesService', () => {
  let service: CurrenciesService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [CurrenciesService],
    }).compile();

    service = module.get<CurrenciesService>(CurrenciesService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('check inputCurrency is invalid', () => {
    const checkCountryFunction = () => {
      service.fetchExchangeRate('T', 'USD');
    };

    expect(checkCountryFunction).toThrow(BadRequestException);
    expect(checkCountryFunction).toThrow(COUNTRY_IS_INVALID);
  });

  it('check targetCurrency is invalid', () => {
    const checkCountryFunction = () => {
      service.fetchExchangeRate('TWD', 'U');
    };

    expect(checkCountryFunction).toThrow(BadRequestException);
    expect(checkCountryFunction).toThrow(COUNTRY_IS_INVALID);
  });

  it('fetchExchangeRate success', () => {
    const exchangeRatePO: ExchangeRatePO = { targetRate: 111.801 };

    expect(service.fetchExchangeRate('USD', 'JPY')).toEqual(exchangeRatePO);
  });

  it('exchangeValue success', () => {
    expect(service.exchangeValue(111.801, 222)).toEqual(24819.822);
  });

  it('exchangeValue fail', () => {
    const testFunction = () => {
      service.exchangeValue(0.0003, 0.002);
    };

    expect(testFunction).toThrow(BadRequestException);
    expect(testFunction).toThrow(VALUE_IS_INVALID);
  });

  it('addThousandsSeparatorAndRoundingDigits success', () => {
    expect(service.standardizeValue(31555125.8456)).toEqual('31,555,125.85');
    expect(service.standardizeValue(31555125.8446)).toEqual('31,555,125.84');
  });
});


src/currencies/currencies.service.ts
import { BadRequestException, Injectable } from '@nestjs/common';
import { ExchangeCurrencyQueryDTO } from './dto/currency.dto';
import { ExchangeRatePO } from './po/currency.po';
import { plainToClass } from 'class-transformer';
import { ExchangeCurrencyAO } from './ao/currency.ao';
import { ExceptionConst } from '../const/exception.const';
import { evaluate } from 'mathjs';
const { COUNTRY_IS_INVALID, VALUE_IS_INVALID } = ExceptionConst;

@Injectable()
export class CurrenciesService {
  exchangeCurrency(dto: ExchangeCurrencyQueryDTO): ExchangeCurrencyAO {
    const { inputCurrency, targetCurrency, value } = dto;

    const exchangeRatePO = this.fetchExchangeRate(
      inputCurrency,
      targetCurrency,
    );

    const exchangedValue = this.exchangeValue(exchangeRatePO.targetRate, value);

    const valueWithCurrencySeparators = this.standardizeValue(exchangedValue);

    return plainToClass(
      ExchangeCurrencyAO,
      {
        targetRate: exchangeRatePO.targetRate,
        targetCurrency,
        value: valueWithCurrencySeparators,
      },
      { excludeExtraneousValues: true },
    );
  }

  fetchExchangeRate(
    inputCurrency: string,
    targetCurrency: string,
  ): ExchangeRatePO {
    // fetch exchange rate from ThirdParty API
    const currencies = {
      TWD: {
        TWD: 1,
        JPY: 3.669,
        USD: 0.03281,
      },
      JPY: {
        TWD: 0.26956,
        JPY: 1,
        USD: 0.00885,
      },
      USD: {
        TWD: 30.444,
        JPY: 111.801,
        USD: 1,
      },
    };

    if (!currencies[inputCurrency]) {
      throw new BadRequestException(COUNTRY_IS_INVALID);
    }

    if (!currencies[inputCurrency][targetCurrency]) {
      throw new BadRequestException(COUNTRY_IS_INVALID);
    }

    const targetRate: number = currencies[inputCurrency][targetCurrency];

    return plainToClass(
      ExchangeRatePO,
      { targetRate },
      { excludeExtraneousValues: true },
    );
  }

  exchangeValue(exchangeRate: number, value: number): number {
    const exchangedValue = evaluate(`${exchangeRate} * ${value}`);

    if (!exchangedValue || !(exchangedValue >= 0.005)) {
      throw new BadRequestException(VALUE_IS_INVALID);
    }

    return exchangedValue;
  }

  standardizeValue(value: number): string {
    return value.toLocaleString([], {
      maximumFractionDigits: 2,
      minimumFractionDigits: 2,
    });
  }
}
